_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,133 @@
+---
+layout: post
+published: true
+title: 'Removal of Storage Writes Before Calls to Conditionally Transaction-Terminating Functions'
+date: '2022-09-22'
+author: Solidity Team
+category: Security Alerts
+---
+On September 5, 2022, a bug in Solidity's Yul optimizer was found by differential fuzzing.
+
+The bug was introduced in version 0.8.13 and primarily affects via-IR code generation with enabled optimizer. However, it can also affect optimized legacy code generation with inline assembly blocks involving user-defined functions.
+[Solidity version 0.8.17](https://github.com/ethereum/solidity/releases/tag/v0.8.17), released on September 07, 2022, provides a fix.
+
+We assigned the bug a severity of "medium/high".
+
+
+## Which Contracts are Affected?
+
+A prerequisite for a contract to be affected is the use of ``return(...)`` or ``stop()`` in an inline assembly block (the Solidity ``return`` statement is not affected). Such an inline assembly call will not return (as in continue control flow after the statement), but instead result in an early successful (i.e. non-reverting) termination of the entire transaction.
+
+If this termination of the transaction happens conditionally within a function, i.e. the function can either successfully terminate the transaction or return (as in exit to continue control flow after the call at the call site), the optimizer may incorrectly remove state changes before a call to the function.
+
+This happens in cases, in which the state change before the function could be considered redundant, if the fact that the function can terminate the transaction was ignored.
+
+When using the legacy code generator, the Yul optimizer step responsible for the bug will be run on each inline assembly block in isolation and hence the above conditions have to occur inside one inline assembly block. While still possible, this makes the occurrence of the bug with legacy code generation less likely (see the section of Technical Details below for examples of affected assembly snippets).
+
+However, the new via-IR code generation translates the entire contract to Yul and runs the optimizer on all of it. In this case, the bug can be triggered more easily. The following is a minimal example:
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1; // This write used to be removed by the yul optimizer.
+		g();
+		x = 2;
+	}
+	function g() internal {
+		// The important aspect of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which it terminates the transaction, since the state change above will be missing.
+		if (msg.data.length > 4) return;
```

Comment:
I just now realized that it's `4` because you're checking if there's anything past the signature. It's simple in hindsight but still, maybe we could use something like `msg.value == 0` to make it clearer to users that it's just a dummy condition that does not mean much?

Reply:
Yeah, let's just add a parameter.

Reply:
It could even be `bool _terminateEarly`

Reply:
No one reads the manual. Or comments :P